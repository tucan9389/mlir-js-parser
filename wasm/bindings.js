// Minimal JS wrapper to call the C API from the Emscripten-generated module.
// Assumes the Emscripten build exports the function `mlir_parse_to_string`.

export async function createParserModule(moduleFactory) {
  // moduleFactory is the Emscripten Module factory (e.g., import from generated .js)
  const maybeModule = await moduleFactory();
  const Module = (maybeModule && typeof maybeModule.then === 'function')
    ? await maybeModule
    : maybeModule;
  if (Module && Module.ready && typeof Module.ready.then === 'function') {
    await Module.ready;
  }

  function parseMlir(text) {
    const c_mlir = Module.cwrap('mlir_parse_to_string', 'number', ['string', 'number', 'number', 'number', 'number']);

    // Allocate output and error buffers (start with 64KB each; grow on demand)
    let outCap = 64 * 1024;
    let errCap = 8 * 1024;
    let outPtr = Module._malloc(outCap);
    let errPtr = Module._malloc(errCap);

    try {
      let rc = c_mlir(text, outPtr, outCap, errPtr, errCap);
      if (rc === 0) {
        const outStr = Module.UTF8ToString(outPtr);
        return { ok: true, module: outStr };
      }
      if (rc < 0) {
        const need = -rc;
        Module._free(outPtr);
        outPtr = Module._malloc(need);
        outCap = need;
        rc = c_mlir(text, outPtr, outCap, errPtr, errCap);
        if (rc === 0) {
          const outStr = Module.UTF8ToString(outPtr);
          return { ok: true, module: outStr };
        }
      }
      const errStr = Module.UTF8ToString(errPtr) || 'parse failed';
      return { ok: false, error: errStr };
    } finally {
      Module._free(outPtr);
      Module._free(errPtr);
    }
  }

  return { Module, parseMlir };
}

// UTF8ToString is provided by Emscripten EXPORTED_RUNTIME_METHODS
