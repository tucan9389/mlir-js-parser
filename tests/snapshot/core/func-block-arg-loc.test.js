import { describe, it, expect } from 'vitest';
import { createParserModule } from '../../../wasm/bindings.js';

let ModuleFactory = null;
try {
  const mod = await import('../../../wasm/mlir_parser.js');
  ModuleFactory = mod.default ?? mod;
} catch (_) {
  // no wasm artifact -> tests will be skipped
}

function normalizeLoc(obj) {
  // Make snapshot stable: ensure key order and only keep known loc shape
  if (obj && typeof obj === 'object') {
    if (obj.loc && typeof obj.loc === 'object') {
      const { file, line, column, unknown } = obj.loc;
      obj.loc = unknown ? { unknown: true } : { file, line, column };
    }
    if (Array.isArray(obj.regions)) {
      for (const r of obj.regions || []) {
        for (const b of r.blocks || []) {
          for (const arg of b.arguments || []) normalizeLoc(arg);
          for (const op of b.operations || []) normalizeLoc(op);
        }
      }
    }
  }
  return obj;
}

describe('Snapshot: func.func with block-arg loc', () => {
  it('parses a simple function and captures argument loc', async () => {
    if (!ModuleFactory) return; // skip
    const { parseMlirJson } = await createParserModule(ModuleFactory);
    const src = `module {
      func.func @foo(%x: i32, %y: i32) {
        %0 = arith.addi %x, %y : i32
        func.return
      }
    }`;
    const res = parseMlirJson(src);
    expect(res.ok).toBe(true);
    const json = normalizeLoc(res.json);

    // Quick asserts before snapshot
    expect(json.name).toBe('builtin.module');
    const block0 = json.regions[0].blocks[0];
    // Inside module body: expect one operation (func.func)
    expect(block0.operations.length).toBeGreaterThanOrEqual(1);
    const funcOp = block0.operations.find(o => o.name === 'func.func');
    expect(funcOp).toBeTruthy();
    const entry = funcOp.regions[0].blocks[0];
    expect(entry.arguments.length).toBe(2);
    // Both block args should have loc
    expect(entry.arguments[0].loc).toBeTypeOf('object');
    expect(entry.arguments[1].loc).toBeTypeOf('object');

    // Snapshot entire JSON (normalized) for stability
    expect(json).toMatchSnapshot();
  });
});
