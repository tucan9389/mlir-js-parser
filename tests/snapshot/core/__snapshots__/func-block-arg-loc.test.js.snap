// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot: func.func with block-arg loc > parses a simple function and captures argument loc 1`] = `
{
  "attributes": {},
  "loc": {
    "column": 1,
    "file": "<input>",
    "line": 1,
  },
  "name": "builtin.module",
  "operands": [],
  "regions": [
    {
      "blocks": [
        {
          "arguments": [],
          "operations": [
            {
              "attributes": {
                "function_type": "(i32, i32) -> ()",
                "sym_name": "foo",
              },
              "loc": {
                "column": 7,
                "file": "<input>",
                "line": 2,
              },
              "name": "func.func",
              "operands": [],
              "regions": [
                {
                  "blocks": [
                    {
                      "arguments": [
                        {
                          "loc": {
                            "column": 22,
                            "file": "<input>",
                            "line": 2,
                          },
                          "type": "i32",
                        },
                        {
                          "loc": {
                            "column": 31,
                            "file": "<input>",
                            "line": 2,
                          },
                          "type": "i32",
                        },
                      ],
                      "operations": [
                        {
                          "attributes": {
                            "overflowFlags": "#arith.overflow<none>",
                          },
                          "loc": {
                            "column": 14,
                            "file": "<input>",
                            "line": 3,
                          },
                          "name": "arith.addi",
                          "operands": [
                            {
                              "type": "i32",
                            },
                            {
                              "type": "i32",
                            },
                          ],
                          "regions": [],
                          "results": [
                            {
                              "type": "i32",
                            },
                          ],
                        },
                        {
                          "attributes": {},
                          "loc": {
                            "column": 9,
                            "file": "<input>",
                            "line": 4,
                          },
                          "name": "func.return",
                          "operands": [],
                          "regions": [],
                          "results": [],
                        },
                      ],
                    },
                  ],
                },
              ],
              "results": [],
            },
          ],
        },
      ],
    },
  ],
  "results": [],
}
`;
